File: Semester 4/Medii de proiectare si programare/Laborator/Laborator 3/FestivalMuzicaJava/src/main/java/repository/TicketsDBRepository.java
package repository;

import domain.Show;
import domain.Ticket;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import utils.JdbcUtils;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

public class TicketsDBRepository implements TicketsRepository {

    private final String TABLE_NAME = "tickets";
    private final Logger logger = LogManager.getLogger();
    private final JdbcUtils jdbcUtils;

    public TicketsDBRepository(JdbcUtils jdbcUtils) {
        this.jdbcUtils = jdbcUtils;
    }

    @Override
    public int size() {
        logger.traceEntry();
        int result = 0;

        try (Connection connection = jdbcUtils.getConnection()) {
            PreparedStatement statement = connection.prepareStatement(
                    String.format("SELECT COUNT(*) FROM %s", TABLE_NAME)
            );
            logger.info("Executing: " + statement);

            ResultSet resultSet = statement.executeQuery();
            if (resultSet.next())
                result = resultSet.getInt(1);
        } catch (SQLException e) {
            logger.error(e);
        }

        logger.traceExit(result);
        return result;
    }

    @Override
    public void save(Ticket entity) {
        logger.traceEntry();

        try (Connection connection = jdbcUtils.getConnection()) {
            PreparedStatement statement = connection.prepareStatement(
                    String.format("INSERT INTO %s (client_name, show_id, no_seats) VALUES (?, ?, ?)", TABLE_NAME)
            );

            statement.setString(1, entity.getClientName());
            statement.setLong(2, entity.getShow().getId());
            statement.setInt(3, entity.getNoSeats());
            statement.executeUpdate();
        } catch (SQLException e) {
            logger.error(e);
        }

        logger.traceExit();
    }

    @Override
    public void delete(Long aLong) {
        logger.traceEntry();

        try (Connection connection = jdbcUtils.getConnection()) {
            PreparedStatement statement = connection.prepareStatement(
                    String.format("DELETE FROM %s WHERE id = ?", TABLE_NAME)
            );
            logger.info("Executing: " + statement);

            statement.setLong(1, aLong);
            statement.executeUpdate();
        } catch (SQLException e) {
            logger.error(e);
        }

        logger.traceExit();
    }

    @Override
    public void update(Long aLong, Ticket entity) {
        logger.traceEntry();

        try (Connection connection = jdbcUtils.getConnection()) {
            PreparedStatement statement = connection.prepareStatement(
                    String.format("UPDATE %s SET client_name = ?, show_id = ?, no_seats = ? WHERE id = ?", TABLE_NAME)
            );
            logger.info("Executing: " + statement);

            statement.setString(1, entity.getClientName());
            statement.setLong(2, entity.getShow().getId());
            statement.setInt(3, entity.getNoSeats());
            statement.setLong(4, aLong);
            statement.executeUpdate();
        } catch (SQLException e) {
            logger.error(e);
        }

        logger.traceExit();
    }

    @Override
    public Ticket findOne(Long aLong) {
        logger.traceEntry();
        Ticket result = null;

        try (Connection connection = jdbcUtils.getConnection()) {
            PreparedStatement statement = connection.prepareStatement(
                    String.format("SELECT t.id, t.client_name, t.no_seats, " +
                            "s.id AS show_id, s.artist_name, s.date, s.location, s.remaining_seats " +
                            "FROM %s t JOIN shows s ON t.show_id = s.id WHERE t.id = ?", TABLE_NAME)
            );
            logger.info("Executing: " + statement);

            statement.setLong(1, aLong);
            ResultSet resultSet = statement.executeQuery();

            if (resultSet.next()) {
                result = buildTicketFromJoinResultSet(resultSet);
            }
        } catch (SQLException e) {
            logger.error(e);
        }

        logger.traceExit(result);
        return result;
    }

    @Override
    public Iterable<Ticket> findAll() {
        logger.traceEntry();
        List<Ticket> result = new ArrayList<>();

        try (Connection connection = jdbcUtils.getConnection()) {
            PreparedStatement statement = connection.prepareStatement(
                    String.format("SELECT t.id, t.client_name, t.no_seats, " +
                            "s.id AS show_id, s.artist_name, s.date, s.location, s.remaining_seats " +
                            "FROM %s t JOIN shows s ON t.show_id = s.id", TABLE_NAME)
            );
            logger.info("Executing: " + statement);

            ResultSet resultSet = statement.executeQuery();
            while (resultSet.next()) {
                result.add(buildTicketFromJoinResultSet(resultSet));
            }
        } catch (SQLException e) {
            logger.error(e);
        }

        logger.traceExit(result);
        return result;
    }

    private Ticket buildTicketFromJoinResultSet(ResultSet resultSet) throws SQLException {
        Show show = new Show(
                resultSet.getLong("show_id"),
                resultSet.getString("artist_name"),
                resultSet.getDate("date").toLocalDate(),
                resultSet.getString("location"),
                resultSet.getInt("remaining_seats")
        );

        return new Ticket(
                resultSet.getLong("id"),
                resultSet.getString("client_name"),
                show,
                resultSet.getInt("no_seats")
        );
    }
}

File: Semester 4/Medii de proiectare si programare/Laborator/Laborator 2/FestivalMuzicaCSharp/FestivalMuzicaCSharp/Repository/TicketsDBRepository.cs
using FestivalMuzicaCSharp.Domain;
using FestivalMuzicaCSharp.Utils;
using log4net;

namespace FestivalMuzicaCSharp.Repository;

public class TicketsDBRepository : ITicketsRepository
{
    private const string TableName = "tickets";
    private readonly ILog Log = LogManager.GetLogger(typeof(TicketsDBRepository));
    private IDictionary<string, string> props;

    public TicketsDBRepository(IDictionary<string, string> props)
    {
        this.props = props;
    }

    public int Size()
    {
        Log.InfoFormat("Entering Size");
        var con = DBUtils.getConnection(props);
        var result = 0;

        using (var comm = con.CreateCommand())
        {
            comm.CommandText = $"select count(*) from {TableName}";

            Log.InfoFormat("Executing: {0}", comm.CommandText);

            result = Convert.ToInt32(comm.ExecuteScalar());
        }

        Log.InfoFormat("Exiting Size with value {0}", result);
        return result;
    }

    public void Save(Ticket entity)
    {
        Log.InfoFormat("Entering Save with value {0}", entity);
        var con = DBUtils.getConnection(props);

        using (var comm = con.CreateCommand())
        {
            comm.CommandText = $"insert into {TableName}(client_name, show_id, no_seats) values(@clientName, @showId, @noSeats)";

            var paramClientName = comm.CreateParameter();
            paramClientName.ParameterName = "@clientName";
            paramClientName.Value = entity.ClientName;
            comm.Parameters.Add(paramClientName);

            var paramShowId = comm.CreateParameter();
            paramShowId.ParameterName = "@showId";
            paramShowId.Value = entity.Show.Id;
            comm.Parameters.Add(paramShowId);

            var paramNoSeats = comm.CreateParameter();
            paramNoSeats.ParameterName = "@noSeats";
            paramNoSeats.Value = entity.NoSeats;
            comm.Parameters.Add(paramNoSeats);

            Log.InfoFormat("Executing: {0}", comm.CommandText);

            comm.ExecuteNonQuery();
        }

        Log.InfoFormat("Exiting Save");
    }

    public void Delete(long id)
    {
        Log.InfoFormat("Entering Delete with value {0}", id);
        var con = DBUtils.getConnection(props);

        using (var comm = con.CreateCommand())
        {
            comm.CommandText = $"delete from {TableName} where id = @id";

            var paramId = comm.CreateParameter();
            paramId.ParameterName = "@id";
            paramId.Value = id;
            comm.Parameters.Add(paramId);

            Log.InfoFormat("Executing: {0}", comm.CommandText);

            comm.ExecuteNonQuery();
        }

        Log.InfoFormat("Exiting Delete");
    }

    public void Update(long id, Ticket entity)
    {
        Log.InfoFormat("Entering Update with value {0}", entity);
        var con = DBUtils.getConnection(props);

        using (var comm = con.CreateCommand())
        {
            comm.CommandText = $"update {TableName} set client_name = @clientName, show_id = @showId, no_seats = @noSeats where id = @id";

            var paramClientName = comm.CreateParameter();
            paramClientName.ParameterName = "@clientName";
            paramClientName.Value = entity.ClientName;
            comm.Parameters.Add(paramClientName);

            var paramShowId = comm.CreateParameter();
            paramShowId.ParameterName = "@showId";
            paramShowId.Value = entity.Show.Id;
            comm.Parameters.Add(paramShowId);

            var paramNoSeats = comm.CreateParameter();
            paramNoSeats.ParameterName = "@noSeats";
            paramNoSeats.Value = entity.NoSeats;
            comm.Parameters.Add(paramNoSeats);

            var paramId = comm.CreateParameter();
            paramId.ParameterName = "@id";
            paramId.Value = id;
            comm.Parameters.Add(paramId);

            Log.InfoFormat("Executing: {0}", comm.CommandText);

            comm.ExecuteNonQuery();
        }

        Log.InfoFormat("Exiting Update");
    }

    public Ticket? FindOne(long id)
    {
        Log.InfoFormat("Entering FindOne with value {0}", id);
        var con = DBUtils.getConnection(props);
        Ticket? result = null;

        using (var comm = con.CreateCommand())
        {
            comm.CommandText =
                $"select t.id, t.client_name, t.no_seats, s.id, s.artist_name, s.date, s.location, s.remaining_seats " +
                $"from {TableName} t inner join shows s on t.show_id = s.id where t.id = @id";

            var paramId = comm.CreateParameter();
            paramId.ParameterName = "@id";
            paramId.Value = id;
            comm.Parameters.Add(paramId);

            Log.InfoFormat("Executing: {0}", comm.CommandText);

            using (var dataR = comm.ExecuteReader())
            {
                if (dataR.Read())
                {
                    result = BuildTicketFromJoinRow(dataR);
                }
            }
        }

        Log.InfoFormat("Exiting FindOne with value {0}", result);
        return result;
    }

    public IEnumerable<Ticket> FindAll()
    {
        Log.InfoFormat("Entering FindAll");
        var con = DBUtils.getConnection(props);
        var result = new List<Ticket>();

        using (var comm = con.CreateCommand())
        {
            comm.CommandText =
                $"select t.id, t.client_name, t.no_seats, s.id, s.artist_name, s.date, s.location, s.remaining_seats " +
                $"from {TableName} t inner join shows s on t.show_id = s.id";

            Log.InfoFormat("Executing: {0}", comm.CommandText);

            using (var dataR = comm.ExecuteReader())
            {
                while (dataR.Read())
                {
                    result.Add(BuildTicketFromJoinRow(dataR));
                }
            }
        }

        Log.InfoFormat("Exiting FindAll with value {0}", result);
        return result;
    }

    private static Ticket BuildTicketFromJoinRow(System.Data.IDataReader dataR)
    {
        var ticketId = dataR.GetInt64(0);
        var clientName = dataR.GetString(1);
        var noSeats = dataR.GetInt32(2);

        var showId = dataR.GetInt64(3);
        var artistName = dataR.GetString(4);
        var date = DateOnly.FromDateTime(dataR.GetDateTime(5));
        var location = dataR.GetString(6);
        var remainingSeats = dataR.GetInt32(7);

        var show = new Show(showId, artistName, date, location, remainingSeats);
        return new Ticket(ticketId, clientName, show, noSeats);
    }
}